lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
   void getAnalysisUsage(AnalysisUsage &AU) const override {
     AU.addRequired<BlockFrequencyInfoWrapperPass>();
   }
-
 };
 }
 
 char StructFieldCacheAnalysisPass::ID = 0;
 INITIALIZE_PASS_BEGIN(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
+INITIALIZE_PASS_DEPENDENCY(BlockFrequencyInfoWrapperPass)
 INITIALIZE_PASS_END(StructFieldCacheAnalysisPass, "struct-field-cache-analysis", "Struct Field Cache Analysis", false, false)
 ModulePass *llvm::createStructFieldCacheAnalysisPass() { return new StructFieldCacheAnalysisPass; }
 
-static bool performStructFieldCacheAnalysis(Module &M)
+namespace llvm{
+typedef unsigned FieldNumType;
+
+class StructFieldAccessInfo;
+class StructFieldAccessManager
+{
+  /* This class is used to keep track of all StructFieldAccessInfo objects in the program
+     and make sure only one StructFieldAccessInfo object for each type of struct declared
+     in the program.
+  */
+ public:
+  enum StructDefinitionType{
+    GlobalStruct,
```

Comment:
Is there any requirement on the name of enum? e.g. Do I have to end with "Kind" or "Type"? I checked that LLVM says it needs to be "Kind" as a discriminator for a union, or an indicator of a subclass. But I don't use it for the case.